flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/dsv2/watermark/CountSales.java

```diff
@@ -0,0 +1,281 @@
+     * will typically be advanced to downstream operators in most scenarios. Thus, we create a
+     * WatermarkDeclaration instance that can be used to declare and generate the watermark.
+     */
+    public static final LongWatermarkDeclaration EVENT_TIME_WATERMARK_DECLARATION =
+            WatermarkDeclarations.newBuilder("EVENT_TIME")
+                    .typeLong()
+                    .combineFunctionMin()
+                    .combineWaitForAllChannels(true)
+                    .defaultHandlingStrategyForward()
+                    .build();
+
+    public static void main(String[] args) throws Exception {
+        // parse the parameters
+        final ParameterTool params = ParameterTool.fromArgs(args);
+        final int parallelism = params.getInt("parallelism", 5);
+
+        // obtain execution environment
+        ExecutionEnvironment env = ExecutionEnvironment.getInstance();
+
+        // Create the Order source, the source will declare and generate event time watermarks.
+        NonKeyedPartitionStream<Order> source =
+                env.fromSource(new WrappedSource<>(new OrderSource()), "order source")
+                        .withParallelism(parallelism);
+
+        source
+                // key by product id
+                .keyBy(order -> order.productId)
+                .process(
+                        // handle event time watermark in downstream
+                        new CountSalesProcessFunction())
+                .toSink(new WrappedSink<>(new PrintSink<>()));
+
+        // execute program
+        env.execute("Count Sales");
+    }
+
+    /** Source of Orders. We will declare and generate the event time watermark in this source. */
+    private static class OrderSource extends DataGeneratorSource<Order> {
+
+        public OrderSource() {
+            super(new CustomGeneratorFunction(), 1_000_000L, TypeInformation.of(Order.class));
+        }
+
+        @Override
+        public Set<? extends WatermarkDeclaration> declareWatermarks() {
+            // Declare the event time watermark.
+            return Set.of(EVENT_TIME_WATERMARK_DECLARATION);
+        }
+    }
+
+    /**
+     * Generator function for the Order source. We will generate event time watermark every second
+     * within this generator function.
+     */
+    private static class CustomGeneratorFunction implements GeneratorFunction<Long, Order> {
+
+        private final Random random = new Random();
+
+        private SourceReaderContext readerContext;
+
+        private long previousTimestamp = 0;
+
+        @Override
+        public void open(SourceReaderContext readerContext) throws Exception {
+            this.readerContext = readerContext;
+        }
+
+        @Override
+        public Order map(Long value) throws Exception {
+            long productId = random.nextLong(1, 100);
+            double price = random.nextDouble(1, 1000);
```

Comment:
These random generating apis are introduced in Java 17. There will be compiling errors on Java 11.